README.md
[![crates.io](http://meritbadge.herokuapp.com/sloth)](https://crates.io/crates/sloth)
[![docs.rs](https://docs.rs/sloth/badge.svg)](https://docs.rs/sloth)
[![Build Status](https://travis-ci.org/terrapass/rs-sloth.svg?branch=master)](https://travis-ci.org/terrapass/rs-sloth)

# sloth

This crate provides a generic pointer-like [`Lazy<T, Eval>`](https://docs.rs/sloth/0.2.0/sloth/struct.Lazy.html) struct for lazily initialized values.
It can be used for expensive-to-calculate values to ensure that the evaluation logic runs
only once and only if needed.

For example:
```rust
use sloth::Lazy;

fn get_expensive_string() -> String {
    // do something expensive here to obtain the result,
    // such as read and process file contents
    String::from("some expensive string we got from a file or something")
}

fn get_expensive_number() -> i32 {
    // do something expensive here to calculate the result,
    // such as build a supercomputer and wait 7.5 million years
    42
}

let lazy_string = Lazy::new(get_expensive_string);
let lazy_number = Lazy::new(get_expensive_number);

//...
let must_use_string = true;
//...

if must_use_string {
    println!("Expensive string is: {}", *lazy_string);
    println!("It has length: {}", lazy_string.len());

    // get_expensive_string() has been called only once,
    // get_expensive_number() has not been called
} else {
    println!("Expensive number is: {}", *lazy_number);
    println!("Its square is {}", lazy_number.pow(2));

    // get_expensive_string() has not been called,
    // get_expensive_number() has been called only once
}

```

The evaluated value of a mutable [`Lazy`](https://docs.rs/sloth/0.2.0/sloth/struct.Lazy.html) can be modified:
```rust
use sloth::Lazy;

let mut lazy_vec = Lazy::new(|| vec![2, -5, 6, 0]);

lazy_vec.retain(|n| *n > 0);

assert_eq!(*lazy_vec, vec![2, 6]);
```

[`Lazy`](https://docs.rs/sloth/0.2.0/sloth/struct.Lazy.html) can be consumed and turned into its value via [`unwrap()`](https://docs.rs/sloth/0.2.0/sloth/struct.Lazy.html#method.unwrap):
```rust
use sloth::Lazy;

let lazy_value = Lazy::new(|| "moo");

let output = String::from("a cow goes ") + lazy_value.unwrap();
```


CHANGELOG.md
# `sloth` changelog

## 0.2.0 (2019-09-03)
* Added `Deref<Target = T>`, `DerefMut`, `AsRef<T>`, `AsMut<T>`, `Borrow<T>`,
`BorrowMut<T>` implementations for `Lazy<T, Eval>`.
* Added `unwrap()` method, which consumes `Lazy<T, Eval>` and extracts the evaluated `T` value.
* Deprecated `value_ref()` and `value_mut()` methods in favour of dereference operator `*`, `as_ref()` and `as_mut()`
methods.
* Updated tests and documentation to reflect changes for 0.2.0 release.
* Added `CHANGELOG.md`.

## 0.1.3 (2019-09-01)
* Moved unit-tests and `Lazy` struct implementation to separate files from `lib.rs`.

## 0.1.2 (2019-09-01)
* Updated documentation.

## 0.1.1 (2019-09-01)
* Added Cargo metadata (repository, keywords, categories, badges) to `Cargo.toml`
as well as badges to `README`.

## 0.1.0 (2019-09-01)
* Added `Lazy<T, Eval>` struct, implementing lazily initialized values.
* Added `README.md` and `README.tpl` (for cargo readme) as well as description 
and license to `Cargo.toml`.


src/tests.rs
use super::*;

use std::borrow::{
    Borrow,
    BorrowMut
};

//
// Macros
//

macro_rules! lazy_all_ref_assert_eq {
    ($lazy_value:expr, $expected:expr) => {
        assert_eq!(*$lazy_value, $expected);
        assert_eq!(*$lazy_value.as_ref(), $expected);
        assert_eq!(*$lazy_value.as_mut(), $expected);
        assert_eq!(*$lazy_value.value_ref(), $expected);
        assert_eq!(*$lazy_value.value_mut(), $expected);

        assert!(is_borrow_eq(&$lazy_value, &$expected));
    };
}

//
// Tests
//

#[test]
fn lazy_init_unwrap() {
    let lazy_value = Lazy::new(|| 250);

    assert_eq!(lazy_value.unwrap(), 250);
}

#[test]
fn lazy_init_value() {
    let lazy_value = Lazy::new(|| 125);

    assert_eq!(lazy_value.value(), 125);
}

#[test]
fn lazy_init_value_ref() {
    let lazy_value = Lazy::new(|| 45);

    assert_eq!(*lazy_value.value_ref(), 45);
}

#[test]
fn lazy_init_value_mut() {
    let mut lazy_value = Lazy::new(|| 694);

    assert_eq!(*lazy_value.value_mut(), 694);
}

#[test]
fn lazy_init_deref() {
    let lazy_value = Lazy::new(|| -24);

    assert_eq!(*lazy_value, -24);
}

#[test]
fn lazy_init_deref_mut() {
    use std::ops::DerefMut;

    let mut lazy_value = Lazy::new(|| 955);

    assert_eq!(*lazy_value.deref_mut(), 955);
}

#[test]
fn lazy_init_as_ref() {
    let lazy_value = Lazy::new(|| -23);

    assert_eq!(*lazy_value.as_ref(), -23);
}

#[test]
fn lazy_init_as_mut() {
    let mut lazy_value = Lazy::new(|| 55);

    assert_eq!(*lazy_value.as_mut(), 55);
}

#[test]
fn lazy_init_borrow() {
    let lazy_value = Lazy::new(|| 10);

    assert!(is_borrow_eq(&lazy_value, &10));
}

#[test]
fn lazy_init_borrow_mut() {
    let mut lazy_value = Lazy::new(|| -76);

    let value_mut: &i32 = lazy_value.borrow_mut();

    assert_eq!(*value_mut, -76);
}

#[test]
fn lazy_int_value_retrieval() {
    let mut lazy_value = Lazy::new(|| 5 + 5);

    lazy_all_ref_assert_eq!(lazy_value, 10);
    assert_eq!(lazy_value.value(), 10);
    assert_eq!(lazy_value.unwrap(), 10);
}

#[test]
fn lazy_int_value_modification_value_mut() {
    let mut lazy_value = Lazy::new(|| -1);

    *lazy_value.value_mut() = 42;

    lazy_all_ref_assert_eq!(lazy_value, 42);
    assert_eq!(lazy_value.value(), 42);
    assert_eq!(lazy_value.unwrap(), 42);
}

#[test]
fn lazy_int_value_modification_deref_mut() {
    let mut lazy_value = Lazy::new(|| -1);

    *lazy_value = 42;

    lazy_all_ref_assert_eq!(lazy_value, 42);
    assert_eq!(lazy_value.value(), 42);
    assert_eq!(lazy_value.unwrap(), 42);
}

#[test]
fn lazy_int_value_modification_borrow_mut() {
    let mut lazy_value = Lazy::new(|| -1);

    *lazy_value.borrow_mut() = 42;

    lazy_all_ref_assert_eq!(lazy_value, 42);
    assert_eq!(lazy_value.value(), 42);
    assert_eq!(lazy_value.unwrap(), 42);
}

#[test]
fn lazy_str_value_retrieval() {
    let mut lazy_value = Lazy::new(|| "some str");

    lazy_all_ref_assert_eq!(lazy_value, "some str");
    assert_eq!(lazy_value.value(), "some str");
    assert_eq!(lazy_value.unwrap(), "some str");
}

#[test]
fn lazy_str_value_modification_value_mut() {
    let mut lazy_value = Lazy::new(|| "initial str");

    *lazy_value.value_mut() = "new str";

    lazy_all_ref_assert_eq!(lazy_value, "new str");
    assert_eq!(lazy_value.value(), "new str");
    assert_eq!(lazy_value.unwrap(), "new str");
}

#[test]
fn lazy_str_value_modification_deref_mut() {
    let mut lazy_value = Lazy::new(|| "initial str");

    *lazy_value = "new str";

    lazy_all_ref_assert_eq!(lazy_value, "new str");
    assert_eq!(lazy_value.value(), "new str");
    assert_eq!(lazy_value.unwrap(), "new str");
}

#[test]
fn lazy_str_value_modification_borrow_mut() {
    let mut lazy_value = Lazy::new(|| "initial str");

    *lazy_value.borrow_mut() = "new str";

    lazy_all_ref_assert_eq!(lazy_value, "new str");
    assert_eq!(lazy_value.value(), "new str");
    assert_eq!(lazy_value.unwrap(), "new str");
}

#[test]
fn lazy_string_value_retrieval() {
    let mut lazy_value = Lazy::new(|| "some string".to_string());

    lazy_all_ref_assert_eq!(lazy_value, "some string".to_string());
    assert_eq!(lazy_value.unwrap(), "some string".to_string());
}

#[test]
fn lazy_string_value_modification_value_mut() {
    let mut lazy_value = Lazy::new(|| "initial string".to_string());

    *lazy_value.value_mut() = "new string".to_string();

    lazy_all_ref_assert_eq!(lazy_value, "new string".to_string());
    assert_eq!(lazy_value.unwrap(), "new string".to_string());
}

#[test]
fn lazy_string_value_modification_deref_mut() {
    let mut lazy_value = Lazy::new(|| "initial string".to_string());

    *lazy_value = "new string".to_string();

    lazy_all_ref_assert_eq!(lazy_value, "new string".to_string());
    assert_eq!(lazy_value.unwrap(), "new string");
}

#[test]
fn lazy_string_value_modification_borrow_mut() {
    let mut lazy_value = Lazy::new(|| "initial string".to_string());

    *lazy_value.borrow_mut() = "new string".to_string();

    lazy_all_ref_assert_eq!(lazy_value, "new string".to_string());
    assert_eq!(lazy_value.unwrap(), "new string");
}

#[test]
#[allow(unused_variables)]
fn lazy_evaluator_never_called_if_unused() {
    let mut evaluator_call_count = 0;

    let lazy_value = Lazy::new(|| {
        evaluator_call_count += 1;
        25
    });

    assert_eq!(evaluator_call_count, 0);
}

#[test]
#[allow(unused_must_use)]
fn lazy_evaluator_called_once() {
    let mut evaluator_call_count = 0;

    let mut lazy_value = Lazy::new(|| {
        evaluator_call_count += 1;
        150
    });

    *lazy_value;
    lazy_value.as_ref();
    *lazy_value.as_mut() = 150;
    let _: i32 = *lazy_value.borrow();
    *lazy_value.borrow_mut() = 250;
    lazy_value.value();
    lazy_value.value();
    *lazy_value.value_mut() = 200;
    lazy_value.value();

    assert_eq!(evaluator_call_count, 1);
}

#[test]
#[allow(unused_must_use)]
fn lazy_value_drop_if_used() {
    let mut was_value_dropped = false;

    {
        let lazy_value = Lazy::new(|| SomethingDroppable{was_dropped: &mut was_value_dropped});

        lazy_value.value_ref();
    }

    assert!(was_value_dropped);
}

#[test]
#[allow(unused_variables)]
fn lazy_value_no_drop_if_unused() {
    let mut was_value_dropped = false;

    {
        let lazy_value = Lazy::new(|| SomethingDroppable{was_dropped: &mut was_value_dropped});
    }

    assert!(!was_value_dropped);
}

#[test]
fn lazy_value_no_drop_on_unwrap() {
    let mut was_value_dropped = false;

    {
        let lazy_value = Lazy::new(|| SomethingDroppable{was_dropped: &mut was_value_dropped});

        let value = lazy_value.unwrap();

        assert!(!*value.was_dropped);
    }

    assert!(was_value_dropped);
}

//
// Service
//

fn is_borrow_eq<T, Q>(borrowed: &Q, expected: &T) -> bool
    where   T: PartialEq,
            Q: Borrow<T>
{
    borrowed.borrow() == expected.borrow()
}

//
// Service types
//

struct SomethingDroppable<'a> {
    was_dropped: &'a mut bool
}

impl Drop for SomethingDroppable<'_> {
    fn drop(&mut self) {
        assert!(!*self.was_dropped, "was_dropped must initially be false");

        *self.was_dropped = true;
    }
}


src/lazy.rs
use std::cell::{
    Cell,
    RefCell,
    Ref,
    RefMut
};
use std::ops::{
    Deref,
    DerefMut
};
use std::borrow::{
    Borrow,
    BorrowMut
};

//
// Constants
//

const EXPECT_VALUE_CELL_INITIALIZED:  &str = "option in value_cell must be initialized at this point";
const EXPECT_EVALUATOR_STILL_PRESENT: &str = "evaluator must still be present at this point";
const EXPECT_VALUE_CELL_PTR_NOT_NULL: &str = "value_cell as ptr must not be null";

//
// Interface
//

//
// struct Lazy<T, Eval>: Deref<Target = T> + DerefMut + AsRef<T> + Borrow<T> + BorrowMut<T>
//

/// Contains a value of some type `T`, lazily evaluated using a parameterless
/// function or a closure (`FnOnce() -> T`) passed to [`Lazy::new()`](struct.Lazy.html#method.new).
/// 
/// This type provides pointer-like interface to the evaluation result,
/// implementing [`Deref`](https://doc.rust-lang.org/nightly/core/ops/deref/trait.Deref.html),
/// [`AsRef`](https://doc.rust-lang.org/nightly/core/convert/trait.AsRef.html),
/// [`Borrow`](https://doc.rust-lang.org/nightly/core/borrow/trait.Borrow.html)
/// and their `Mut` counterparts.
/// 
/// The result will be evaluated the first time it is accessed via any of `Lazy`'s methods.
/// 
/// # Accessing evaluated value
/// 
/// A `Lazy` value can be acessed in any of the following ways:
/// ```
/// use sloth::Lazy;
/// 
/// let mut lazy_value = Lazy::new(|| 2019);
/// 
/// let value: i32 = *lazy_value;
/// let value_ref: &i32 = lazy_value.as_ref();
/// let value_mut: &mut i32 = lazy_value.as_mut();
/// 
/// let value_copy: i32 = lazy_value.value(); // only for types implementing Copy
/// 
/// let value: i32 = lazy_value.unwrap(); // consumes lazy_value
/// ```
/// 
/// Due to [`Deref` coercion](https://doc.rust-lang.org/std/ops/trait.Deref.html#more-on-deref-coercion)
/// `T`'s methods may be called directly on [`Lazy<T, Eval>`](struct.Lazy.html), while references to `Lazy`
/// are coerced to references to `T`:
/// ```
/// use sloth::Lazy;
/// 
/// fn print_string_len(string: &String) {
///     println!("{} has length {}", string, string.len());
/// }
/// 
/// let mut lazy_string = Lazy::new(|| String::from("lorem "));
/// 
/// lazy_string.push_str("ipsum"); // can call T's methods
/// 
/// print_string_len(&lazy_string); // can pass as &T function param
///
/// ```
/// 
/// # Laziness
/// 
/// The evaluator function will not be called more than once,
/// regardless of how many times the value is accessed:
/// ```
/// use sloth::Lazy;
/// 
/// let mut evaluator_called_times = 0;
/// 
/// let lazy_value = Lazy::new(|| {
///     evaluator_called_times += 1;
///     25
/// });
/// 
/// assert_eq!(*lazy_value, 25);
/// 
/// let another_value = *lazy_value + *lazy_value;
/// 
/// assert_eq!(evaluator_called_times, 1);
/// ```
/// 
/// If a `Lazy` value is never dereferenced and none of its methods are called,
/// its evaluator function will not be invoked at all.
pub struct Lazy<T, Eval>
    where Eval: FnOnce() -> T
{
    evaluator_cell: Cell<Option<Eval>>,
    value_cell:     RefCell<Option<T>>
}

//
// Trait impls
//

impl<T, Eval> Deref for Lazy<T, Eval>
    where Eval: FnOnce() -> T
{
    type Target = T;

    /// Immutable dereference, allowing access to the contained value.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    fn deref(&self) -> &T {
        self.as_ref_impl()
    }
}

impl<T, Eval> DerefMut for Lazy<T, Eval>
    where Eval: FnOnce() -> T
{
    /// Mutable dereference, allowing access to the contained value.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    fn deref_mut(&mut self) -> &mut T {
        self.as_mut_impl()
    }
}

impl<T, Eval> AsRef<T> for Lazy<T, Eval>
    where Eval: FnOnce() -> T
{
    /// Immutably borrows the evaluation result.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    fn as_ref(&self) -> &T {
        self.as_ref_impl()
    }
}

impl<T, Eval> AsMut<T> for Lazy<T, Eval>
    where Eval: FnOnce() -> T
{
    /// Mutably borrows the evaluation result.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    fn as_mut(&mut self) -> &mut T {
        self.as_mut_impl()
    }
}

impl<T, Eval> Borrow<T> for Lazy<T, Eval>
    where Eval: FnOnce() -> T
{
    /// Immutably borrows the evaluation result.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    fn borrow(&self) -> &T {
        self.as_ref_impl()
    }
}

impl<T, Eval> BorrowMut<T> for Lazy<T, Eval>
    where Eval: FnOnce() -> T
{
    /// Mutably borrows the evaluation result.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    fn borrow_mut(&mut self) -> &mut T {
        self.as_mut_impl()
    }
}

//
// Methods
//

impl<T, Eval> Lazy<T, Eval>
    where Eval: FnOnce() -> T
{
    //
    // Interface
    //

    /// Constructs a lazy `T` instance, whose value, if needed, will later be
    /// obtained from `evaluator` and cached.
    /// 
    /// `evaluator` will be invoked only the first time this instance
    /// is dereferenced or one of its methods is invoked.
    pub fn new(evaluator: Eval) -> Self {
        Self{
            evaluator_cell: Cell::new(Some(evaluator)),
            value_cell:     RefCell::new(None)
        }
    }

    /// Immutably borrows the evaluation result.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    /// 
    /// **`value_ref()` will be removed in sloth 0.3.0. [`as_ref()` or immutable * dereference](struct.Lazy.html#implementations) should be used instead.**
    #[must_use]
    #[deprecated(since = "0.2.0", note = "will be removed in sloth 0.3.0; please use as_ref() or * deref operator instead")]
    pub fn value_ref(&self) -> Ref<'_, T> {
        self.init_once();

        // Returns a Ref to the T instance contained within Option<T> referenced by value_cell
        Ref::map(
            self.value_cell.borrow(),
            |value_option| {
                value_option.as_ref().expect(EXPECT_VALUE_CELL_INITIALIZED)
            }
        )
    }

    /// Mutably borrows the evaluation result.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    /// 
    /// **`value_mut()` will be removed in sloth 0.3.0. [`as_mut()` or mutable * dereference](struct.Lazy.html#implementations) should be used instead.**
    #[must_use]
    #[deprecated(since = "0.2.0", note = "will be removed in sloth 0.3.0; please use as_mut() or * deref operator instead")]
    pub fn value_mut(&mut self) -> RefMut<'_, T> {
        self.init_once();

        // Returns a RefMut to the T instance contained within Option<T> referenced by value_cell_mut
        RefMut::map(
            self.value_cell.borrow_mut(),
            |value_option| {
                value_option.as_mut().expect(EXPECT_VALUE_CELL_INITIALIZED)
            }
        )
    }

    /// Consumes this [`Lazy<T, Eval>`](struct.Lazy.html) instance and extracts the evaluation result value.
    ///
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    #[must_use]
    pub fn unwrap(self) -> T {
        self.init_once();

        self.value_cell.replace(None).expect(EXPECT_VALUE_CELL_INITIALIZED)
    }

    //
    // Service
    //

    fn as_ref_impl(&self) -> &T {
        self.init_once();

        unsafe {
            self.value_cell
                .as_ptr()
                .as_ref()
                .expect(EXPECT_VALUE_CELL_PTR_NOT_NULL)
                .as_ref()
                .expect(EXPECT_VALUE_CELL_INITIALIZED)
        }
    }

    fn as_mut_impl(&mut self) -> &mut T {
        self.init_once();

        unsafe {
            self.value_cell
                .as_ptr()
                .as_mut()
                .expect(EXPECT_VALUE_CELL_PTR_NOT_NULL)
                .as_mut()
                .expect(EXPECT_VALUE_CELL_INITIALIZED)
        }
    }

    fn init_once(&self) {
        if self.value_cell.borrow().is_none() {
            *self.value_cell.borrow_mut() = Some(self.evaluate());
        }
    }

    fn evaluate(&self) -> T {
        let evaluator = self.evaluator_cell
            .take()
            .expect(EXPECT_EVALUATOR_STILL_PRESENT);

        evaluator()
    }
}

impl<T, Eval> Lazy<T, Eval>
    where T:    Copy,
          Eval: FnOnce() -> T
{
    /// Returns a copy of the evaluation result.
    /// 
    /// This will invoke evaluator function if none of the methods
    /// or `*` deref operator were previously used.
    #[must_use]
    pub fn value(&self) -> T {
        self.init_once();

        self.value_cell.borrow().expect(EXPECT_VALUE_CELL_INITIALIZED)
    }
}


src/lib.rs
//! This crate provides a generic pointer-like [`Lazy<T, Eval>`](struct.Lazy.html) struct for lazily initialized values.
//! It can be used for expensive-to-calculate values to ensure that the evaluation logic runs
//! only once and only if needed.
//! 
//! For example:
//! ```
//! use sloth::Lazy;
//! 
//! fn get_expensive_string() -> String {
//!     // do something expensive here to obtain the result,
//!     // such as read and process file contents
//!     String::from("some expensive string we got from a file or something")
//! }
//! 
//! fn get_expensive_number() -> i32 {
//!     // do something expensive here to calculate the result,
//!     // such as build a supercomputer and wait 7.5 million years
//!     42
//! }
//! 
//! let lazy_string = Lazy::new(get_expensive_string);
//! let lazy_number = Lazy::new(get_expensive_number);
//! 
//! //...
//! let must_use_string = true;
//! //...
//! 
//! if must_use_string {
//!     println!("Expensive string is: {}", *lazy_string);
//!     println!("It has length: {}", lazy_string.len());
//!
//!     // get_expensive_string() has been called only once,
//!     // get_expensive_number() has not been called
//! } else {
//!     println!("Expensive number is: {}", *lazy_number);
//!     println!("Its square is {}", lazy_number.pow(2));
//! 
//!     // get_expensive_string() has not been called,
//!     // get_expensive_number() has been called only once
//! }
//! 
//! ```
//! 
//! The evaluated value of a mutable [`Lazy`](struct.Lazy.html) can be modified:
//! ```
//! use sloth::Lazy;
//! 
//! let mut lazy_vec = Lazy::new(|| vec![2, -5, 6, 0]);
//! 
//! lazy_vec.retain(|n| *n > 0);
//! 
//! assert_eq!(*lazy_vec, vec![2, 6]);
//! ```
//! 
//! [`Lazy`](struct.Lazy.html) can be consumed and turned into its value via [`unwrap()`](struct.Lazy.html#method.unwrap):
//! ```
//! use sloth::Lazy;
//! 
//! let lazy_value = Lazy::new(|| "moo");
//! 
//! let output = String::from("a cow goes ") + lazy_value.unwrap();
//! ```

//
// Modules
//

mod lazy;

// Unit tests
#[cfg(test)]
#[allow(deprecated)]
mod tests;

//
// Exports
//

pub use lazy::Lazy;
